packages/web/app/src/lib/fetch-json.ts

```diff
@@ -0,0 +1,134 @@
+import { z } from 'zod';
+import { Kit } from './kit';
+
+export async function fetchJson<schema extends undefined | z.ZodType = undefined>(
+  url: string,
+  requestInit?: RequestInit,
+  schema?: schema,
+): Promise<
+  (schema extends z.ZodType ? z.infer<schema> : Kit.Json.Value) | FetchJsonErrors.FetchJsonErrors
+> {
+  const response = await fetch(url, requestInit)
+    // @see https://developer.mozilla.org/en-US/docs/Web/API/Window/fetch#exceptions
+    .catch(Kit.oneOf(error => error instanceof TypeError || error instanceof DOMException));
+
+  if (response instanceof TypeError) {
+    return new FetchJsonErrors.FetchJsonRequestTypeError({ requestInit }, response);
+  }
+
+  if (response instanceof DOMException) {
+    return new FetchJsonErrors.FetchJsonRequestNetworkError({}, response);
+  }
+
+  const json = await response
+    .json()
+    .then(value => value as Kit.Json.Value)
+    // @see https://developer.mozilla.org/en-US/docs/Web/API/Response/json#exceptions
+    .catch(
+      Kit.oneOf(
+        error =>
+          error instanceof SyntaxError ||
+          error instanceof TypeError ||
+          error instanceof DOMException,
+      ),
+    );
+
+  if (json instanceof DOMException) {
+    return new FetchJsonErrors.FetchJsonRequestNetworkError({}, json);
+  }
+
+  if (json instanceof TypeError) {
+    return new FetchJsonErrors.FetchJsonResponseTypeError({ response }, json);
+  }
+
+  if (json instanceof SyntaxError) {
+    return new FetchJsonErrors.FetchJsonResponseSyntaxError({ response }, json);
+  }
+
+  if (schema) {
+    const result = schema.safeParse(json);
+    if (!result.success) {
+      return new FetchJsonErrors.FetchJsonResponseSchemaError(
+        { response, json, schema },
+        result.error,
+      );
+    }
+    return result.data as any; // z.infer<Exclude<schema, undefined>>;
+  }
+
+  return json as any; // Kit.Json.Value;
```

Comment:
nit: prefer `unknown` to force us to use validation or cast

Reply:
Since this is in the return position of an implementation it would not work and it does not matter, TS just doesn't make it possible to make implementations type safe. At best we could cast it to something matching the return type but its more maintenance for little gain and no increased type safety since it is a cast either way.